File: receptor.py
def verificar_y_corregir_hamming(trama: str) -> tuple:
    bits = list(map(int, trama))

    # Posiciones de los bits de paridad en Hamming
    posiciones_paridad = [0, 1, 3, 7] 

    error_pos = 0
    for i, p in enumerate(posiciones_paridad):
        suma = 0
        for j in range(len(bits)):
            if j != p and ((j + 1) & (2 ** i)) != 0:
                suma ^= bits[j]
        if suma != bits[p]:
            error_pos += 2 ** i

    if error_pos > 0:
        print(f"Error detectado en la posición {error_pos}. Corrigiendo...")
        bits[error_pos - 1] ^= 1  # Corrige el bit
        corregido = True
    else:
        corregido = False

    # Extraer los bits
    datos = [bits[i] for i in range(len(bits)) if i not in posiciones_paridad]
    mensaje_original = ''.join(map(str, datos))

    return mensaje_original, corregido, error_pos if corregido else None

def procesar_receptor(archivo):
    with open(archivo, 'r') as f:
        lineas = f.readlines()

    for i, linea in enumerate(lineas):
        trama = linea.strip()
        if not trama:
            continue
        mensaje, corregido, posicion = verificar_y_corregir_hamming(trama)
        print(f"Trama {i+1}: {trama}")
        if corregido:
            print(f"- Error corregido en bit {posicion}. Mensaje original: {mensaje}\n")
        else:
            print(f"- Sin errores. Mensaje original: {mensaje}\n")


procesar_receptor("receptor.txt")

File: receptor.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace std;

// Función para verificar y corregir errores en una trama Hamming 
tuple<string, bool, int> verificar_y_corregir_hamming(const string& trama) {
    vector<int> bits;
    for (char c : trama) {
        bits.push_back(c - '0');
    }

    // Posiciones de bits
    vector<int> posiciones_paridad = {0, 1, 3, 7};

    int error_pos = 0;
    for (int i = 0; i < posiciones_paridad.size(); ++i) {
        int p = posiciones_paridad[i];
        int suma = 0;
        for (int j = 0; j < bits.size(); ++j) {
            if (j != p && ((j + 1) & (1 << i))) {
                suma ^= bits[j];
            }
        }
        if (suma != bits[p]) {
            error_pos += (1 << i);
        }
    }

    bool corregido = false;
    if (error_pos > 0 && error_pos <= bits.size()) {
        cout << "Error detectado en la posición " << error_pos << ". Corrigiendo..." << endl;
        bits[error_pos - 1] ^= 1;
        corregido = true;
    }

    // Extraer los bits de datos
    string mensaje_original = "";
    for (int i = 0; i < bits.size(); ++i) {
        if (find(posiciones_paridad.begin(), posiciones_paridad.end(), i) == posiciones_paridad.end()) {
            mensaje_original += to_string(bits[i]);
        }
    }

    return {mensaje_original, corregido, error_pos};
}

void procesar_receptor(const string& archivo) {
    ifstream file(archivo);
    if (!file) {
        cerr << "No se pudo abrir el archivo receptor.txt" << endl;
        return;
    }

    string linea;
    int num_trama = 1;

    while (getline(file, linea)) {
        if (linea.empty()) continue;

        auto [mensaje, corregido, posicion] = verificar_y_corregir_hamming(linea);

        cout << "Trama " << num_trama << ": " << linea << endl;
        if (corregido) {
            cout << "- Error corregido en bit " << posicion << ". Mensaje original: " << mensaje << endl << endl;
        } else {
            cout << "- Sin errores. Mensaje original: " << mensaje << endl << endl;
        }

        num_trama++;
    }

    file.close();
}

int main() {
    procesar_receptor("receptor.txt");
    return 0;
}

// Ejecutar:
// g++ receptor.cpp -std=c++17 -o receptor
// ./receptor
